File: mergeSort.py
def mergeSort(my_List):
    if len(my_List) > 1:
        mid = len(my_List) // 2
        left = my_List[:mid]
        right = my_List[mid:]

        mergeSort(left)
        mergeSort(right)

        i = 0
        j = 0
        k = 0

        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                my_List[k] = left[i]
                i = i + 1
            else:
                my_List[k] = right[j]
                j = j + 1
            k = k + 1
        
        while i < len(left):
            my_List[k] = left[i]
            i = i + 1
            k = k + 1

        while j < len(right):
            my_List[k] = right[j]
            j = j + 1
            k = k + 1
    
    return my_List

my_List = [44, 16, 24, 83, 75, 25, 9, 22, 10]
mergeSortedList = mergeSort(my_List)
print(mergeSortedList)

File: Sorting Algorithms/JavaScript/MergeSort.js
// Merge sort algorithm is JavaScript
// Out-of-place algorithm - requires growing amount of memory to process the input into output.
// Stable - Two elements with equal values will appear in the same order in the sorted output as they appear in the unsorted output.

function mergeSort(arr) {
    // Best case.
    if (arr.length <= 1) return arr;

    let mid = Math.floor(arr.length / 2)

    let left = arr.slice(0, mid)
    let right = arr.slice(mid)

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right);
}

function merge(left, right) {
    // Create an empty array:
    sortedArray = [];
    while (left.length && right.length) {
        // Insert the smallest item in the array
        if (left[0] < right[0]) {
            sortedArray.push(left.shift())
        } else {
            sortedArray.push(right.shift())
        }
    }

    return [...sortedArray, ...left, ...right]
}

array = [40,1,5,28,17,34,2]
console.log(mergeSort(array));


